enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+
+CLI tooling will be implemented to help users who wish to disable the CCO to perform their initial installations, update credentials prior to upgrade for release images available for update, and reconcile credentials against current CredentialsRequests in the cluster on an on-going basis. Users must take this manual step to clear the Upgradable=False condition if credentials have changed in any available update.
+
+Cloud credential operator reconcile code will be refactored to be a little more friendly to use as a vendored library, for use in the CLI.
+
+Note that when we say CCO is disabled, it is still running, just not minting credentials, and does not have access to the powerful "root" credentials for the cloud provider account.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+Core of this proposal will be CLI tooling which vendors code from the credentials operator and allows for single use execution of the reconcile loop. Provising CLI tooling allows users to bypass the security concerns of storing powerful cloud credentials in the cluster and instead keep them entirely locally.
+
+### Upgradable False Condition
+
+A field will be added to the cloud credential operator ConfigMap (possibly becoming a separate API type in another enhancement) storing a list of Updates (from ClusterVersion.AvailableUpdates) that have been reconciled, confirming that we have minted credentials and it is safe to upgrade to this release.
+
+```yaml
+  reconciledAvailableUpdates:
+  - version: 4.4.1
+    image: quay.io/openshift-release-dev/ocp-release:4.4.1-x86_64
+  - version: 4.4.2
+    image: quay.io/openshift-release-dev/ocp-release:4.4.2-x86_64
+```
+
+Entries will be added to this list by the CLI tooling below, however an administrator could do this by hand if desired.
+
+If the CCO is enabled per normal best practice install, this field will be ignored, and should be absent.
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
+
+### Upgrade Reconciliation
+
+```bash
+$ oc adm credentials reconcile --available-updates
+```
+
+This command will iterate all currently available updates in ClusterVersion.AvailableUpdates. For each given release image we will extract each CredentialsRequest and use the local users environment variables/credentials file to compare and adjust permissions as necessary.
```

Comment:
is there a recommendation for a user that does not want to have us mint credentials but is also running disconnected, so available updates is empty?

Reply:
> ...but is also running disconnected...

If we get this enhancement before setting up CVO preflights (which still need their own enhancement), this helper should allow the caller to pass in a release image pullspec to work around this "I have no local Cincinnati" situation.

also nit: "iterate all" -> "iterate over all" and "users" -> "user's"

Reply:
I was not aware of this use case, we may need some more details, how does a disconnected user initiate an upgrade? I assume it's somehow just specifying a specific release image to upgrade to? That does pose a problem as I'm not sure how we'd get the CCO to clear it's condition if we don't know the release. Feels like the proposed upgrade pre-flight checks I'm learning about here may be able to help.